### ISSUE TITLE
Feature Request: C++ gradient for SoftmaxCrossEntropyWithLogits

### ISSUE THREAD

**Issue by bpiel** [tensorflow/tensorflow#12686]
Implement the gradient for SoftmaxCrossEntropyWithLogits in c++ so that it is available for TF_AddGradients.

This is the python code that I believe would need to be ported:
https://github.com/tensorflow/tensorflow/blob/4b2fb49fd7578afe7e289936f347af581b5bdab1/tensorflow/python/ops/nn_grad.py#L409

**Comment by skye** [comment]
@bpiel you're working on this right?

@suharshs FYI

**Comment by bpiel** [comment]
@skye Not currently. If I start, I will post here. If someone else starts first, hopefully they'll post here.

**Comment by facaiy** [comment]
Hi, I'm really interested in the work because it is a little challenge for me. #12391 of @bpiel seems like a perfect example, right? I could be more confident if more detailed guide are given or someone would like shepherd it.

**Comment by suharshs** [comment]
@facaiy Happy to review your PR. :) 
@kbsriram may also be interested in reviewing/guiding.

**Comment by facaiy** [comment]
Thanks, @suharshs .

I have read the code in python side,
```python
@ops.RegisterGradient("SoftmaxCrossEntropyWithLogits")
def _SoftmaxCrossEntropyWithLogitsGrad(op, grad_loss, grad_grad):
```

If I understand correctly, the arguments `grad_loss` and `grad_grad` in python side will be packed in `grad_inputs` in C++ side, right?

Namely, the outline of code in C++ should like:
```c++
Status SoftmaxCrossEntropyWithLogitsGrad(const Scope& scope, const Operation& op,
                                         const std::vector<Output>& grad_inputs,
                                         std::vector<Output>* grad_outputs) {
  auto grad_loss = grad_inputs[0];
  auto grad_grad = grad_inputs[1];

  // migrate the implementation in python to here.
  // code...........
  // auto grad = .....

  grad_outputs->push_back(grad);
  grad_outputs->push_back(null);
  return scope.status();
}
REGISTER_GRADIENT_OP("SoftmaxCrossEntropyWithLogitsGrad", SoftmaxCrossEntropyWithLogitsGrad);
```

Please correct me if I'm wrong.

**Comment by kbsriram** [comment]
@facaiy the outline of the code looks fine, one change would be to use
```c++
grad_outputs->push_back(NoGradient())
```
rather than `null` to indicate gradients should not be propagated.

One thing you may run into is that that python has various utilities (e.g. [tensor_util.constant_value()](https://github.com/tensorflow/tensorflow/tree/77b5f6a956c61e6ada9d4a6c61892f9bd2464fdb/tensorflow/python/framework/tensor_util.py#L695)) that don't yet have equivalents available in C++. Should you find yourself writing with a long stack of utilities, one strategy might be to start with gradients that have with fewer dependencies. Some options might be functions like [Erf](https://github.com/tensorflow/tensorflow/blob/77b5f6a956c61e6ada9d4a6c61892f9bd2464fdb/tensorflow/python/ops/math_grad.py#L450) and [LGamma](https://github.com/tensorflow/tensorflow/blob/77b5f6a956c61e6ada9d4a6c61892f9bd2464fdb/tensorflow/python/ops/math_grad.py#L471) in `math_ops`. You can look at c++ implementations like [Tanh](https://github.com/tensorflow/tensorflow/blob/8624ecc9e827a40f9b514ff6b8ed925390ca79cc/tensorflow/cc/gradients/math_grad.cc#L191) to see how to structure the code and tests.

**Comment by skye** [comment]
re: @kbsriram's point about missing Python utilities, for SoftmaxCrossEntropyWithLogitsGrad, you can skip the tensor_util.constant_value() call and leave a TODO to implement it in C++. This is OK in this case since the IsZero() check is an optimization anyway (I think...).

**Comment by facaiy** [comment]
@kbsriram @skye Thanks for your enlightening and detailed guide. I agree that it's better to start with easier task at first. Hence, I'd like to work on `LGamma` and `Erf` as suggested.

If someone else is interested in `SoftmaxCrossEntropyWithLogitsGrad` as well, please let us know, and just go head and take over it.

**Comment by pbanavara** [comment]
I would like to take over this feature. What level of C++ expertise is needed. I'm a total C++ noob but am seeing this as an opportunity to learn the language as well as make a meaningful contribution. I was just checking at the python code and this line caught my eye
`softmax = nn_ops.softmax(logits)`
I presume the C++ equivalent of nn_ops.softmax(logits) is implemented somewhere - I couldn't find out where. Is this correct ? Can someone please point out where to find the equivalent c++ implementation for nn_ops.softmax.

**Comment by facaiy** [comment]
@pbanavara [`tensorflow::ops::Softmax`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/softmax) ?

**Comment by skye** [comment]
Hi @pbanavara, @facaiy is correct. We have functions for every op in both Python and C++. I would look at the C++ API documentation (https://www.tensorflow.org/api_guides/cc/guide, https://www.tensorflow.org/api_docs/cc/group/core) and existing gradient functions to get a feel for what they look like. Also make sure you can successfully build TensorFlow before you start writing code. Maybe I'm overly familiar with the code, but I don't think gradient functions should be too hard to start with. If SoftmaxCrossEntropyWithLogitsGrad proves to be too complicated you can also try a simpler one. Good luck, and looking forward to a PR!

**Comment by pbanavara** [comment]
@facaiy @skye Thank you for the pointers. I will start working on this and will ping you guys once I have the tests working.

**Comment by suharshs** [comment]
Also, if you want instructions for adding the c++ gradient functions see here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/gradients/README.md

**Comment by bpiel** [comment]
@pbanavara This is coming up for me again. How's it going?

**Comment by pbanavara** [comment]
@bpiel Going slow. I was down with a minor cervical spine injury. So couldn't type much. Will keep you posted. Sorry I couldn't update earlier.

**Comment by bpiel** [comment]
@pbanavara Very sorry to hear about your injury. I hope you're doing better. Updates appreciated. thanks!

**Comment by pbanavara** [comment]
Hey guys - I got around to making some progress. Need some help and apologies for super noob questions. This is what I have done to port the Python code

```
Status SoftmaxCrossEntropyWithLogitsGrad(const Scope& scope,
                                          const Operation& op,
                                          const std::vector<Output>&
                                          grad_inputs,
                                          std::vector<Output>* grad_outputs) {
  // Softmax gradient with cross entropy logits function
  // We multiply the backprop for cost with the gradients - op.output[1]
  // There is no gradient for labels

  auto softmaxGrad = op.output(1);
  auto gradLoss = grad_inputs[0];
  auto gradGrad = grad_inputs[1];

  auto grad = Mul(scope, gradLoss, softmaxGrad);

  // TODO Check if the grad is not zero

  auto logits = op.input(0);
  auto softmax = ops::Softmax(scope, logits);

  auto prod = ops::MatMul(scope, gradGrad, softmax);
  auto squeezeProd = ops::Squeeze(scope, prod);
  auto fProd = Sub(scope, gradGrad, squeezeProd);
  grad = Add(scope, grad, fProd);
  grad_outputs->push_back(grad);
  return scope.status();
}
```
I get an error at the line

`grad = Add(scope, grad, fProd);`

tensorflow/cc/gradients/nn_grad.cc:74:8: error: no viable overloaded '='

I have totally forgotten about operator overloading. Have I done some fundamental mistake in the code above or do I have to write a operator overloader for =. 

I don't think I can do something like 
`grad = grad + fProd;`

**Comment by suharshs** [comment]
`Output` (the return type of the op calls) doesn't support assignment (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/framework/ops.h#L84).

Instead of assigning to `grad` which was used for `Mul`, if you assign a new variable (`auto new_var = Add(...)`) things should work. (but pick a better name than `new_var` :) )

**Comment by skye** [comment]
Suharsh's suggestion is correct (create a new variable), but it's not because of Output (you can assign Outputs). Mul and Add are both classes, so when you call them, you're calling their constructors. See https://www.tensorflow.org/api_guides/cc/guide#operation_constructors. So you're creating a Mul variable, then trying to assign an Add to it. You need to create a new Add variable instead.

And if you're wondering, no, Mul and Add (and all the other generated op classes) are not all subclasses of the same base class. So you might be further wondering, how is it legal to copy Mul and Add objects into vector<Output> grad_outputs? We clearly don't have an overloaded Output constructor for every generated op. The answer is that the generated op classes that have a single Output use an implicit Output conversion operator (see http://en.cppreference.com/w/cpp/language/cast_operator).

Also, definitely not a super noob question, no worries :)

**Comment by suharshs** [comment]
Oh i see, thanks for the correction @skye

**Comment by pbanavara** [comment]
Thank you @suharshs and @Skye. Adding a new variable did the job. Follow up question - Can you guys  please tell me what's the equivalent of eager/Context in C++ and where can I find the related API docs. I tried in vain searching through the documentation tree.

**Comment by skye** [comment]
There is no such concept in C++, so you can ignore it.

**Comment by pbanavara** [comment]
Cool, ok, I've got the build and test complete. Test fails so am looking into that. If I can't figure it out by tomorrow will ask you guys again. Thank you for all the help.

**Comment by pbanavara** [comment]
Hey guys sorry I couldn't get the test to pass. Help needed. I just copied the test from the Softmax test and made changes. This is what I have

```
TEST_F(NNGradTest, SoftmaxCrossEntropyWithLogitsGrad) {
  TensorShape shape({5,3});
  auto x = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
  auto l = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
  auto y = SoftmaxCrossEntropyWithLogits(scope_, x, l).backprop;
  RunTest(x, shape, y, shape);
}
```
Here are my questions:
I get a compilation error if I just use 
`auto y = SoftmaxCrossEntropyWithLogits(scope_, x, l)`
So I made an educated guess to use the backprop - is this wrong ?
Is there any such thing as enabling debug flag for tests. In the log I just have a failure message. 

Thanks again for all your help.

**Comment by skye** [comment]
What's the failure message?

SoftmaxCrossEntropyWithLogits has two outputs, loss and backprops: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/nn_ops.cc#L2002

RunTest() calls ComputeGradientError(), so I suspect you somehow need to incorporate both outputs in the gradient test. I'm not sure though. Maybe @suharshs knows more.

**Comment by pbanavara** [comment]
@skye Am getting a segmentation fault. I tried both on Ubuntu and Mac, same result.

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
Running main() from test_main.cc
[==========] Running 12 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 12 tests from NNGradTest
[ RUN      ] NNGradTest.SoftmaxGrad
2017-11-05 06:53:43.521514: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
[       OK ] NNGradTest.SoftmaxGrad (163 ms)
[ RUN      ] NNGradTest.SoftmaxCrossEntropyWithLogitsGrad
Segmentation fault (core dumped)
```

**Comment by skye** [comment]
Ah. Have you used gdb before? I suggest getting a stacktrace, and you can also poke around and find out more about what's causing the error. To use gdb:
1. Build the test in debug mode by using `bazel build -c dbg tensorflow/cc/gradients_nn_grad_test`
2. Open the test binary in gdb: `gdb bazel-bin/tensorflow/cc/gradients_nn_grad_test`
3. Once you get the gdb prompt, run the test in gdb: `(gdb) r`
4. It should automatically break when it hits the segfault. Once this happens, use `bt` to print the stacktrace. You can also inspect variables, etc. I suggest looking online for a basic tutorial if you haven't used gdb before.

Hope that helps, feel free to post anything you find here and I can try to help you more.

**Comment by suharshs** [comment]
Apologies for the delayed response, you don't need to use the `.backprop` in your test. You just write the usage of the op as normal. The gradient checker will add gradients and verify that they match the numerical gradient. 

Here is an example of another test in that same file that does this. The only thing you need to make sure of is that the x shape and y shape match what SoftMaxWithCrossEntropyLogits does, which i believe you have correct.

```
TEST_F(NNGradTest, Conv2DGrad) {
  TensorShape shape({1, 2, 2, 1});
  auto x = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
  Tensor filter = test::AsTensor<float>({0.5f}, {1, 1, 1, 1});
  const std::vector<int> strides{1, 1, 1, 1};
  auto y = Conv2D(scope_, x, filter, strides, "SAME");
  RunTest(x, shape, y, shape);
}
```

**Comment by suharshs** [comment]
You will also need to use the last override of RunTest which can handle multiple outputs and inputs.

```
void RunTest(const OutputList& xs, const std::vector<TensorShape>& x_shapes,
           const OutputList& ys, const std::vector<TensorShape>& y_shapes) 
```

**Comment by pbanavara** [comment]
Hey @suharshs and @skye Thanks and sorry for the delayed response. I haven't gone cold on this - still working. Trying to run the correct RunTest variant. Will keep you posted here soon.

**Comment by pbanavara** [comment]
Ok did some amateurish stuff ... 

```
  OutputList inputs = OutputList();
  std::vector<TensorShape> shapes = std::vector<TensorShape>();
  OutputList outputs = OutputList();
  inputs.push_back(x);
  shapes.push_back(shape);
  auto y = SoftmaxCrossEntropyWithLogits(scope_, x, l);
  outputs.push_back(y);
  RunTest(inputs, shapes, outputs, shapes);

```
I get this error 

`candidate function not viable: no known conversion from 'tensorflow::ops::SoftmaxCrossEntropyWithLogits' to 'const value_type' (aka 'const tensorflow::Output') for 1st argument`

At the line where I am adding the outputs to the OutputList vector. Since the RunTest variant requires an OutputList I did it this way. Perhaps there is another way to do this ? 
SoftmaxCrossEntropyWithLogits call returns an Output as per the documentation. How it returns multiple outputs is not clear to me. Can you guys please help out with that ?

**Comment by skye** [comment]
It's a little confusing, but SoftmaxCrossEntropyWithLogits has two outputs, 'loss' and 'backprop'. These are public properties of the returned SoftmaxCrossEntropyWithLogits instance (see https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/softmax-cross-entropy-with-logits), so can be accessed in your code via `y.loss` and `y.backprop`. Both of these should be added the `outputs` list I think.

**Comment by pbanavara** [comment]
Thanks @skye That solved the error. I had tried this before but had called loss and backdrop as methods :). I overlooked the fact that they are just attributes.  
Now the output shapes are not matching. I am trying to fix that. The error messages are quite hard to interpret.
```
[ RUN      ] NNGradTest.SoftmaxCrossEntropyWithLogitsGrad
tensorflow/cc/gradients/nn_grad_test.cc:55: Failure
      Expected: ::tensorflow::Status::OK()
      Which is: OK
To be equal to: ((ComputeGradientError<float, float, float>( scope_, xs, x_shapes, ys, y_shapes, &max_error)))
      Which is: Invalid argument: Dimensions must be equal, but are 3 and 5 for 'MatMul' (op: 'MatMul') with input shapes: [5,3], [5,3].
```
So clearly one of the matrices need to be of shape [3,5] but which one. So I am just trying all combinations  in the inputs and outputs.

**Comment by skye** [comment]
Could this be an actual bug in your SoftmaxCrossEntropyWithLogits function? If you can't find the bug, post what you have so far in a PR and I can try to take a look.

**Comment by pbanavara** [comment]
Quite possibly it’s a big in my code. Am investigating the same and will
post a PR soon

**Comment by pbanavara** [comment]
@skye Have submitted a PR. I will keep debugging the test case fail issue in the interim. Please let me know if you find some obvious mistake.

**Comment by skye** [comment]
Hey @pbanavara, just a heads up that I'm pretty slammed this week so I can't look into this right now :( Sorry to keep you hanging. Maybe someone else cc'd can take a look, otherwise I'll try to make some time next week.

**Comment by pbanavara** [comment]
Hey @skye no worries at all. please comment when you can.
